source/slang/slang-check-decl.cpp

```diff
@@ -8943,6 +8980,146 @@ namespace Slang
+
+            // Reset everything
+            backingWidth = 0;
+            totalWidth = 0;
+            groupInfo.clear();
+        };
+        for(; memberIndex < structDecl->members.getCount(); ++memberIndex)
+        {
+            const auto& m = structDecl->members[memberIndex];
+
+            // We can trivially skip any non-property decls
+            const auto v = as<PropertyDecl>(m);
+            if(!v)
+            {
+                // If this is a non-bitfield member then finish the current group
+                if(as<VarDecl>(m))
+                    dispatchSomeBitPackedMembers();
+                continue;
+            }
+
+            const auto bfm = m->findModifier<BitFieldModifier>();
+            // If there isn't a bit field modifier, then dispatch the
+            // current group and continue
+            if(!bfm)
+            {
+                dispatchSomeBitPackedMembers();
+                continue;
+            }
+
+            // Verify that this makes sense as a bitfield
+            const auto t = v->type.type->getCanonicalType();
+            SLANG_ASSERT(t);
+            const auto b = as<BasicExpressionType>(t);
+            if(!b)
+            {
+                getSink()->diagnose(v->loc, Diagnostics::bitFieldNonIntegral, t);
+                continue;
+            }
+            const auto baseType = b->getBaseType();
+            const bool isIntegerType = isIntegerBaseType(baseType);
+            if(!isIntegerType)
+            {
+                getSink()->diagnose(v->loc, Diagnostics::bitFieldNonIntegral, t);
+                continue;
+            }
+
+            // The bit width of this member, and the member type width
+            const auto thisFieldWidth = bfm->width;
+            const auto thisFieldTypeWidth = getTypeBitSize(b);
+            SLANG_ASSERT(thisFieldTypeWidth != 0);
+            if(thisFieldWidth > thisFieldTypeWidth)
+            {
+                getSink()->diagnose(
+                    v->loc,
+                    Diagnostics::bitFieldTooWide,
+                    thisFieldWidth,
+                    t,
+                    thisFieldTypeWidth
+                );
+                // Not much we can do with this field, just ignore it
+                continue;
+            }
+
+            // At this point we're sure that we have a bit field,
+            // update our bit packing state
+
+            // If there's a 0 width type, dispatch the current group
+            if(thisFieldWidth == 0)
+                dispatchSomeBitPackedMembers();
+
+            // If this member wouldn't fit into the current group, dispatch
+            // everything so far;
+            if(totalWidth + thisFieldWidth > std::max(thisFieldTypeWidth, backingWidth))
+                dispatchSomeBitPackedMembers();
+
+            // Add this member to the group,
+            // Grow the backing width if necessary
+            backingWidth = std::max(thisFieldTypeWidth, backingWidth);
```

Comment:
For consistency with C++, I think the behavior should be if the backing with of the currently open backing field is different from the type of this field, then we start a new field.

Reply:
I think that this is implementation defined for C and C++, and at least gcc and clang are happy to push multiple types together into one field

Reply:
https://github.com/shader-slang/slang/issues/3646